File: python_sdk/src/reality_capture/specifications/constraints.py
from pydantic import BaseModel, Field, UUID4
from typing import Optional
from enum import Enum


class ConstraintType(str, Enum):
    MESH = "Mesh"
    POLYGON = "Polygon"


class ConstraintToAdd(BaseModel):
    constraint_path: str = Field(alias="constraintPath",
                                 description="Path to constraint file (bucket path in cloud, local path on-premise)")
    crs: str = Field(description="Coordinate reference system")
    type: Optional[ConstraintType] = Field(None, description="Type of the constraint")
    resolution: Optional[float] = Field(None, description="Resolution of the constraint")
    texture_path: Optional[str] = Field(None, alias="texturePath",
                                        description="Path to texture file (bucket path in cloud, local path on-premise)")
    texture_size: Optional[int] = Field(None, alias="textureSize", description="Size of the texture")
    fill_color: Optional[str] = Field(None, alias="fillColor", description="Fill color for the constraint")
    name: Optional[str] = Field(None, description="Name of the constraint")
    description: Optional[str] = Field(None, description="Description of the constraint")


class ConstraintInfo(ConstraintToAdd):
    id: UUID4 = Field(description="Constraint unique id")
    surfaces: list[str] = Field(description="List of meshes")
    crs_surfaces: str = Field(alias="crsSurfaces", description="Coordinate Reference System of surfaces")


class ConstraintsInfo(BaseModel):
    constraints: list[ConstraintInfo] = Field(description="Constraints information")


class ConstraintsInputs(BaseModel):
    modeling_reference: str = Field(alias="modelingReference", description="Modeling reference to update.")
    constraints_to_delete: Optional[list[UUID4]] = Field(None, alias="constraintsToDelete",
                                                         description="IDs of constraints to delete")
    constraints_to_add: Optional[list[ConstraintToAdd]] = Field(None, alias="constraintsToAdd",
                                                                description="Constraints to add")
    crs_data: Optional[str] = Field(default=None,
                                    description="Path to CRS data file (bucket path in cloud, local path on-premise).",
                                    alias="crsData")


class ConstraintsOutputs(BaseModel):
    added_constraints_info: str = Field(alias="addedConstraintsInfo",
                                        description="Path to added ConstraintsInfo file (bucket path in cloud, local path on-premise)",)


class ConstraintsOutputsCreate(Enum):
    ADDED_CONSTRAINTS_INFO = "addedConstraintsInfo"


class ConstraintsSpecifications(BaseModel):
    inputs: ConstraintsInputs = Field(description="Inputs")
    outputs: ConstraintsOutputs = Field(description="Outputs")


class ConstraintsSpecificationsCreate(BaseModel):
    inputs: ConstraintsInputs = Field(description="Inputs")
    outputs: list[ConstraintsOutputsCreate] = Field(description="Outputs")


class ConstraintsCost(BaseModel):
    surface: float = Field(description="Surface in squared meters of the constraints to add", ge=0)

File: typescript/packages/reality-capture/src/specifications/constraints.ts
import { z } from "zod";

export enum ConstraintType {
  MESH = "Mesh",
  POLYGON = "Polygon",
}

export const ConstraintToAddSchema = z.object({
  constraintPath: z.string().regex(/^bkt:.+/, {
    message: "Path in the bucket to the constraint file must start with 'bkt:'",
  }),
  crs: z.string().describe("Coordinate reference system"),
  type: z.nativeEnum(ConstraintType).optional().describe("Type of the constraint"),
  resolution: z.number().optional().describe("Resolution of the constraint"),
  texturePath: z.string().regex(/^bkt:.+/, {
    message: "Path in the bucket to the texture file must start with 'bkt:'",
  }).optional(),
  textureSize: z.number().int().optional().describe("Size of the texture"),
  fillColor: z.string().optional().describe("Fill color for the constraint"),
  name: z.string().optional().describe("Name of the constraint"),
  description: z.string().optional().describe("Description of the constraint"),
});
export type ConstraintToAdd = z.infer<typeof ConstraintToAddSchema>;

export const ConstraintInfoSchema = ConstraintToAddSchema.extend({
  id: z.string().uuid().describe("Constraint unique id"),
  surfaces: z.array(z.string()).describe("List of meshes"),
  crsSurfaces: z.string().describe("Coordinate Reference System of surfaces"),
});
export type ConstraintInfo = z.infer<typeof ConstraintInfoSchema>;

export const ConstraintsInfoSchema = z.object({
  constraints: z.array(ConstraintInfoSchema).describe("Constraints information"),
});
export type ConstraintsInfo = z.infer<typeof ConstraintsInfoSchema>;

export const ConstraintsInputsSchema = z.object({
  modelingReference: z.string().describe("Modeling reference to update."),
  constraintsToDelete: z.array(z.string().uuid()).optional().describe("IDs of constraints to delete"),
  constraintsToAdd: z.array(ConstraintToAddSchema).optional().describe("Constraints to add"),
  crsData: z.string().regex(/^bkt:.+/).optional().describe("Path in the bucket for CRS data."),
});
export type ConstraintsInputs = z.infer<typeof ConstraintsInputsSchema>;

export const ConstraintsOutputsSchema = z.object({
  addedConstraintsInfo: z.string().regex(/^bkt:.+/, {
    message: "Path in the bucket for added ConstraintsInfo must start with 'bkt:'",
  }).describe("Path in the bucket for added ConstraintsInfo"),
});
export type ConstraintsOutputs = z.infer<typeof ConstraintsOutputsSchema>;

export enum ConstraintsOutputsCreate {
  ADDED_CONSTRAINTS_INFO = "addedConstraintsInfo",
}

export const ConstraintsSpecificationsSchema = z.object({
  inputs: ConstraintsInputsSchema.describe("Inputs"),
  outputs: ConstraintsOutputsSchema.describe("Outputs"),
});
export type ConstraintsSpecifications = z.infer<typeof ConstraintsSpecificationsSchema>;

export const ConstraintsSpecificationsCreateSchema = z.object({
  inputs: ConstraintsInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(ConstraintsOutputsCreate)).describe("Outputs"),
});
export type ConstraintsSpecificationsCreate = z.infer<typeof ConstraintsSpecificationsCreateSchema>;

export const ConstraintsCostSchema = z.object({
  surface: z.number().min(0, "Surface must be >= 0").describe("Surface in squared meters of the constraints to add"),
});
export type ConstraintsCost = z.infer<typeof ConstraintsCostSchema>;